### PR TITLE
Fixed merging circles at the same time causing a crash

### LINKED ISSUE #19922: Occasional crash when merging hit objects
I've tried to understand how this can happen but just can't see it.

The `blueprintMap` pre-check should guarantee that the blueprint is still added as a drawable in the container where the `ChangeChildDepth` is being called, but that doesn't seem to be the case.

Is expiry happening elsewhere or am I missing something?

As a side-note, because the merge operation is calling `BeginChange` on the `IEditorChangeHandler` and not `EditorBeatmap`, the mutation of hitobjects is not completely batching (only from the perspective of the undo/redo state). I'm not sure this is an issue, but worth noting. It is being done a similar way in many other usages.

Sentry Issue: [OSU-596](https://sentry.ppy.sh/organizations/ppy/issues/6138/?referrer=github_integration)

```
System.InvalidOperationException: Can not change depth of drawable which is not contained within this CompositeDrawable.
  ?, in void CompositeDrawable.ChangeInternalChildDepth(Drawable child, float newDepth)
  ?, in void EditorBlueprintContainer.OnBlueprintDeselected(SelectionBlueprint<HitObject> blueprint)
  ?, in void BlueprintContainer<T>.RemoveBlueprintFor(T item)
  ?, in void EditorBeatmap.UpdateState()
  ?, in void TransactionalCommitComponent.EndChange()
...
(16 additional frame(s) were not displayed)

An unhandled error has occurred.
```

### PR DESCRIPTION
**Pull request by OliBomby**
closes #19922 

- [x] Depends on https://github.com/ppy/osu/pull/19982

We might want to make a new issue for the underlying issue? If there are two circles at the same time and both selected and then you remove the first circle (somehow without deleting the second circle aswell), then the game crashes. Not sure if this is intented because you currently cant cause this behaviour in the editor.

### CODE REVIEW

File: osu.Game.Rulesets.Osu/Edit/OsuSelectionHandler.cs [hunk 1]
```diff
@@ -363,6 +363,7 @@ private void mergeSelection()
                 return;
 
             ChangeHandler?.BeginChange();
+            EditorBeatmap.BeginChange();
```
Comment (bdach) [inline osu.Game.Rulesets.Osu/Edit/OsuSelectionHandler.cs:366]: It shouldn't be required to call both of these. `EditorBeatmap.BeginChange()` implies the `ChangeHandler` call automatically.

https://github.com/ppy/osu/blob/3e246caf05ade38bffdc436fdc4e306d1c01dd2e/osu.Game/Screens/Edit/EditorChangeHandler.cs#L58-L60

Same goes for `EndChange()`, obviously.

**Review by bdach — CHANGES_REQUESTED** [review]

**Comment by bdach** [comment]
> We might want to make a new issue for the underlying issue? If there are two circles at the same time and both selected and then you remove the first circle (somehow without deleting the second circle aswell), then the game crashes

This is probably a failure stemming from [`HitObjectOrderedSelectionContainer`'s custom comparer](https://github.com/ppy/osu/blob/master/osu.Game/Screens/Edit/Compose/Components/HitObjectOrderedSelectionContainer.cs), which is causing `CompositeDrawable.ChangeInternalChildDepth()` to fail to find the blueprint because of the internal binary search failing to locate the element (as there are two elements with equal keys in this case). And yes it's probably something we want fixed in the long run.

File: osu.Game.Rulesets.Osu/Edit/OsuSelectionHandler.cs (hunk 1 above)
Comment (OliBomby) [reply osu.Game.Rulesets.Osu/Edit/OsuSelectionHandler.cs:366]: ah I didn't know that

Comment (peppy) [reply osu.Game.Rulesets.Osu/Edit/OsuSelectionHandler.cs]: Is there a reason the change from `ChangeHandler` to `EditorBeatmap` was preferred here? It has very nuanced different behaviour which shouldn't really matter too much, but curious if that was a consideration here.

(I really don't think both of these should exist, and we should only expose this in one way..)

Comment (peppy) [reply osu.Game.Rulesets.Osu/Edit/OsuSelectionHandler.cs]: Okay, so it actually fixes the whole flow in this case. Welp.

**Review by peppy — APPROVED** [review]
Let's go with this..

**Comment by peppy** [comment]
Going to go with this as-is, tracking the weirdness of this fix at https://github.com/ppy/osu/issues/20040.

**Review by bdach — APPROVED** [review]

**Pull request merged**